File: ocp-python/src/ocp_agent/storage.py
"""
OCPStorage - Local storage for API caching and session persistence.

Provides optional local storage for OCP agents to cache API specifications
and persist session contexts across application restarts.
"""

import json
import os
from pathlib import Path
from typing import Optional, List, Dict, Any
from datetime import datetime, timezone, timedelta

from .context import AgentContext
from .schema_discovery import OCPAPISpec


class OCPStorage:
    """
    Local storage manager for OCP API cache and session persistence.
    
    Storage directory structure:
        ~/.ocp/
        ├── cache/
        │   └── apis/
        │       ├── github.json
        │       ├── stripe.json
        │       └── ...
        └── sessions/
            ├── vscode-chat-abc123.json
            ├── cli-session-xyz789.json
            └── ...
    
    Design principles:
    - Per-file storage for surgical reads/writes
    - Fail-safe: storage errors don't break functionality
    - Consumer-agnostic: doesn't manage consumer config
    """
    
    def __init__(self, base_path: Optional[Path] = None):
        """
        Initialize storage with base directory.
        
        Args:
            base_path: Base storage directory (defaults to ~/.ocp/)
        """
        if base_path is None:
            base_path = Path.home() / ".ocp"
        
        self.base_path = Path(base_path)
        self.cache_dir = self.base_path / "cache" / "apis"
        self.sessions_dir = self.base_path / "sessions"
        
        self._ensure_dirs()
    
    def _ensure_dirs(self) -> None:
        """Create storage directories if they don't exist."""
        try:
            self.cache_dir.mkdir(parents=True, exist_ok=True)
            self.sessions_dir.mkdir(parents=True, exist_ok=True)
        except OSError as e:
            # Non-fatal: log but don't raise
            print(f"Warning: Failed to create storage directories: {e}")
    
    # ==================== API Caching ====================
    
    def cache_api(
        self, 
        name: str, 
        spec: OCPAPISpec, 
        metadata: Optional[Dict[str, Any]] = None
    ) -> bool:
        """
        Cache an API specification locally.
        
        Args:
            name: API name (used as filename)
            spec: API specification to cache
            metadata: Optional metadata (source, cached_at, etc.)
            
        Returns:
            True if cached successfully, False on error
        """
        try:
            cache_file = self.cache_dir / f"{name}.json"
            
            cache_data = {
                "api_name": name,
                "title": spec.title,
                "version": spec.version,
                "base_url": spec.base_url,
                "cached_at": datetime.now(timezone.utc).isoformat(),
                "source": metadata.get("source", "unknown") if metadata else "unknown",
                "raw_spec": spec.raw_spec,
                "tools": [
                    {
                        "name": tool.name,
                        "method": tool.method,
                        "path": tool.path,
                        "description": tool.description,
                        "parameters": tool.parameters,
                        "response_schema": tool.response_schema,
                        "operation_id": tool.operation_id,
                        "tags": tool.tags
                    }
                    for tool in spec.tools
                ]
            }
            
            if metadata:
                cache_data.update(metadata)
            
            with open(cache_file, 'w', encoding='utf-8') as f:
                json.dump(cache_data, f, indent=2, ensure_ascii=False)
            
            return True
            
        except (OSError, IOError, ValueError) as e:
            print(f"Warning: Failed to cache API '{name}': {e}")
            return False
    
    def get_cached_api(
        self, 
        name: str, 
        max_age_days: Optional[int] = None
    ) -> Optional[OCPAPISpec]:
        """
        Retrieve cached API specification.
        
        Args:
            name: API name to retrieve
            max_age_days: Maximum age in days (None = no expiration)
            
        Returns:
            Cached API spec or None if not found/expired
        """
        try:
            cache_file = self.cache_dir / f"{name}.json"
            
            if not cache_file.exists():
                return None
            
            with open(cache_file, 'r', encoding='utf-8') as f:
                cache_data = json.load(f)
            
            # Check expiration if max_age_days is set
            if max_age_days is not None:
                cached_at = datetime.fromisoformat(
                    cache_data["cached_at"].replace("Z", "+00:00")
                )
                age = datetime.now(timezone.utc) - cached_at
                
                if age > timedelta(days=max_age_days):
                    return None  # Expired
            
            # Reconstruct OCPAPISpec from cache
            # Import here to avoid circular dependency
            from .schema_discovery import OCPTool
            
            tools = [
                OCPTool(
                    name=t["name"],
                    method=t["method"],
                    path=t["path"],
                    description=t.get("description", ""),
                    parameters=t.get("parameters", {}),
                    response_schema=t.get("response_schema", {}),
                    operation_id=t.get("operation_id"),
                    tags=t.get("tags")
                )
                for t in cache_data.get("tools", [])
            ]
            
            api_spec = OCPAPISpec(
                title=cache_data["title"],
                version=cache_data["version"],
                base_url=cache_data["base_url"],
                description=cache_data.get("description", ""),
                raw_spec=cache_data["raw_spec"],
                tools=tools
            )
            
            return api_spec
            
        except (OSError, IOError, ValueError, KeyError) as e:
            print(f"Warning: Failed to load cached API '{name}': {e}")
            return None
    
    def search_cache(self, query: str) -> List[Dict[str, Any]]:
        """
        Search cached APIs by name or description.
        
        Args:
            query: Search query (case-insensitive)
            
        Returns:
            List of matching API metadata
        """
        results = []
        query_lower = query.lower()
        
        try:
            if not self.cache_dir.exists():
                return []
            
            for cache_file in self.cache_dir.glob("*.json"):
                try:
                    with open(cache_file, 'r', encoding='utf-8') as f:
                        cache_data = json.load(f)
                    
                    # Search in name, title, and tool descriptions
                    if (query_lower in cache_data.get("api_name", "").lower() or
                        query_lower in cache_data.get("title", "").lower() or
                        any(query_lower in tool.get("description", "").lower() 
                            for tool in cache_data.get("tools", []))):
                        
                        results.append({
                            "name": cache_data.get("api_name"),
                            "title": cache_data.get("title"),
                            "version": cache_data.get("version"),
                            "base_url": cache_data.get("base_url"),
                            "cached_at": cache_data.get("cached_at"),
                            "tool_count": len(cache_data.get("tools", []))
                        })
                        
                except (OSError, IOError, ValueError):
                    continue
            
        except OSError:
            pass
        
        return results
    
    def list_cached_apis(self) -> List[str]:
        """
        List all cached API names.
        
        Returns:
            List of API names
        """
        try:
            if not self.cache_dir.exists():
                return []
            
            return [f.stem for f in self.cache_dir.glob("*.json")]
            
        except OSError:
            return []
    
    def clear_cache(self, name: Optional[str] = None) -> bool:
        """
        Clear API cache.
        
        Args:
            name: Specific API to clear (None = clear all)
            
        Returns:
            True if cleared successfully
        """
        try:
            if name:
                # Clear specific API
                cache_file = self.cache_dir / f"{name}.json"
                if cache_file.exists():
                    cache_file.unlink()
            else:
                # Clear all cache files
                for cache_file in self.cache_dir.glob("*.json"):
                    cache_file.unlink()
            
            return True
            
        except OSError as e:
            print(f"Warning: Failed to clear cache: {e}")
            return False
    
    # ==================== Session Persistence ====================
    
    def save_session(self, session_id: str, context: AgentContext) -> bool:
        """
        Save session context to disk.
        
        Args:
            session_id: Unique session identifier
            context: Agent context to save
            
        Returns:
            True if saved successfully
        """
        try:
            session_file = self.sessions_dir / f"{session_id}.json"
            
            session_data = context.to_dict()
            session_data["session_id"] = session_id
            
            with open(session_file, 'w', encoding='utf-8') as f:
                json.dump(session_data, f, indent=2, ensure_ascii=False)
            
            return True
            
        except (OSError, IOError, ValueError) as e:
            print(f"Warning: Failed to save session '{session_id}': {e}")
            return False
    
    def load_session(self, session_id: str) -> Optional[AgentContext]:
        """
        Load session context from disk.
        
        Args:
            session_id: Session identifier to load
            
        Returns:
            Loaded context or None if not found
        """
        try:
            session_file = self.sessions_dir / f"{session_id}.json"
            
            if not session_file.exists():
                return None
            
            with open(session_file, 'r', encoding='utf-8') as f:
                session_data = json.load(f)
            
            # Remove session_id from data (not part of AgentContext)
            session_data.pop("session_id", None)
            
            # Use from_dict to reconstruct context
            context = AgentContext.from_dict(session_data)
            
            return context
            
        except (OSError, IOError, ValueError, KeyError) as e:
            print(f"Warning: Failed to load session '{session_id}': {e}")
            return None
    
    def list_sessions(self, limit: Optional[int] = None) -> List[Dict[str, Any]]:
        """
        List all sessions with metadata.
        
        Args:
            limit: Maximum number of sessions to return (most recent first)
            
        Returns:
            List of session metadata dicts
        """
        sessions = []
        
        try:
            if not self.sessions_dir.exists():
                return []
            
            # Get all session files
            session_files = list(self.sessions_dir.glob("*.json"))
            
            # Sort by modification time (most recent first)
            session_files.sort(key=lambda f: f.stat().st_mtime, reverse=True)
            
            # Apply limit if specified
            if limit:
                session_files = session_files[:limit]
            
            for session_file in session_files:
                try:
                    with open(session_file, 'r', encoding='utf-8') as f:
                        session_data = json.load(f)
                    
                    sessions.append({
                        "id": session_data.get("session_id", session_file.stem),
                        "context_id": session_data.get("context_id"),
                        "agent_type": session_data.get("agent_type"),
                        "user": session_data.get("user"),
                        "workspace": session_data.get("workspace"),
                        "created_at": session_data.get("created_at"),
                        "last_updated": session_data.get("last_updated"),
                        "interaction_count": session_data.get("session", {}).get("interaction_count", 0)
                    })
                    
                except (OSError, IOError, ValueError):
                    continue
            
        except OSError:
            pass
        
        return sessions
    
    def cleanup_sessions(self, keep_recent: int = 50) -> int:
        """
        Remove old sessions, keeping N most recent.
        
        Args:
            keep_recent: Number of recent sessions to keep
            
        Returns:
            Number of sessions removed
        """
        try:
            if not self.sessions_dir.exists():
                return 0
            
            # Get all session files sorted by modification time
            session_files = list(self.sessions_dir.glob("*.json"))
            session_files.sort(key=lambda f: f.stat().st_mtime, reverse=True)
            
            # Remove old sessions beyond keep_recent
            removed = 0
            for session_file in session_files[keep_recent:]:
                try:
                    session_file.unlink()
                    removed += 1
                except OSError:
                    continue
            
            return removed
            
        except OSError as e:
            print(f"Warning: Failed to cleanup sessions: {e}")
            return 0

File: ocp-javascript/src/storage.ts
/**
 * OCPStorage - Local storage for API caching and session persistence.
 *
 * Provides optional local storage for OCP agents to cache API specifications
 * and persist session contexts across application restarts.
 */

import * as fs from 'fs/promises';
import * as fsSync from 'fs';
import * as path from 'path';
import * as os from 'os';
import { AgentContext } from './context.js';
import type { OCPAPISpec, OCPTool } from './schema_discovery.js';

/**
 * Local storage manager for OCP API cache and session persistence.
 *
 * Storage directory structure:
 *     ~/.ocp/
 *     ├── cache/
 *     │   └── apis/
 *     │       ├── github.json
 *     │       ├── stripe.json
 *     │       └── ...
 *     └── sessions/
 *         ├── vscode-chat-abc123.json
 *         ├── cli-session-xyz789.json
 *         └── ...
 *
 * Design principles:
 * - Per-file storage for surgical reads/writes
 * - Fail-safe: storage errors don't break functionality
 * - Consumer-agnostic: doesn't manage consumer config
 */
export class OCPStorage {
    public readonly basePath: string;
    public readonly cacheDir: string;
    public readonly sessionsDir: string;

    /**
     * Initialize storage with base directory.
     *
     * @param basePath - Base storage directory (defaults to ~/.ocp/)
     */
    constructor(basePath?: string) {
        this.basePath = basePath ?? path.join(os.homedir(), '.ocp');
        this.cacheDir = path.join(this.basePath, 'cache', 'apis');
        this.sessionsDir = path.join(this.basePath, 'sessions');

        // Create directories synchronously in constructor
        this.ensureDirsSync();
    }

    /**
     * Create storage directories if they don't exist (synchronous).
     */
    private ensureDirsSync(): void {
        try {
            fsSync.mkdirSync(this.cacheDir, { recursive: true });
            fsSync.mkdirSync(this.sessionsDir, { recursive: true });
        } catch (err) {
            // Non-fatal: log but don't throw
            console.warn(`Warning: Failed to create storage directories: ${err}`);
        }
    }

    // ==================== API Caching ====================

    /**
     * Cache an API specification locally.
     *
     * @param name - API name (used as filename)
     * @param spec - API specification to cache
     * @param metadata - Optional metadata (source, cached_at, etc.)
     * @returns True if cached successfully, false on error
     */
    async cacheApi(
        name: string,
        spec: OCPAPISpec,
        metadata?: Record<string, any>
    ): Promise<boolean> {
        try {
            const cacheFile = path.join(this.cacheDir, `${name}.json`);

            const cacheData = {
                api_name: name,
                title: spec.title,
                version: spec.version,
                base_url: spec.base_url,
                cached_at: new Date().toISOString(),
                source: metadata?.source ?? 'unknown',
                raw_spec: spec.raw_spec,
                tools: spec.tools.map((tool: OCPTool) => ({
                    name: tool.name,
                    method: tool.method,
                    path: tool.path,
                    description: tool.description,
                    parameters: tool.parameters,
                    response_schema: tool.response_schema,
                    operation_id: tool.operation_id,
                    tags: tool.tags
                }))
            };

            if (spec.description) {
                (cacheData as any).description = spec.description;
            }

            await fs.writeFile(
                cacheFile,
                JSON.stringify(cacheData, null, 2),
                'utf-8'
            );

            return true;
        } catch (err) {
            console.warn(`Warning: Failed to cache API '${name}': ${err}`);
            return false;
        }
    }

    /**
     * Retrieve cached API specification.
     *
     * @param name - API name to retrieve
     * @param maxAgeDays - Maximum age in days (0 = no expiration check)
     * @returns Cached API spec or null if not found/expired
     */
    async getCachedApi(
        name: string,
        maxAgeDays: number = 7
    ): Promise<OCPAPISpec | null> {
        try {
            const cacheFile = path.join(this.cacheDir, `${name}.json`);
            const content = await fs.readFile(cacheFile, 'utf-8');
            const data = JSON.parse(content);

            // Check expiration if maxAgeDays > 0
            if (maxAgeDays > 0 && data.cached_at) {
                const cachedAt = new Date(data.cached_at);
                const now = new Date();
                const ageMs = now.getTime() - cachedAt.getTime();
                const ageDays = ageMs / (1000 * 60 * 60 * 24);

                if (ageDays > maxAgeDays) {
                    return null; // Expired
                }
            }

            // Reconstruct OCPAPISpec
            const spec: OCPAPISpec = {
                title: data.title,
                version: data.version,
                base_url: data.base_url,
                description: data.description || '',
                raw_spec: data.raw_spec,
                tools: data.tools.map((tool: any) => ({
                    name: tool.name,
                    method: tool.method,
                    path: tool.path,
                    description: tool.description,
                    parameters: tool.parameters,
                    response_schema: tool.response_schema,
                    operation_id: tool.operation_id,
                    tags: tool.tags
                }))
            };

            return spec;
        } catch (err) {
            // File not found or parse error
            return null;
        }
    }

    /**
     * Search cached APIs by name or description.
     *
     * @param query - Search query (case-insensitive)
     * @returns List of matching API metadata
     */
    async searchCache(query: string): Promise<Array<Record<string, any>>> {
        try {
            const files = await fs.readdir(this.cacheDir);
            const results: Array<Record<string, any>> = [];
            const lowerQuery = query.toLowerCase();

            for (const file of files) {
                if (!file.endsWith('.json')) continue;

                try {
                    const filePath = path.join(this.cacheDir, file);
                    const content = await fs.readFile(filePath, 'utf-8');
                    const data = JSON.parse(content);

                    // Search in name, title, description, and tool descriptions
                    const searchText = [
                        data.api_name || '',
                        data.title || '',
                        data.description || '',
                        ...(data.tools || []).map((t: any) => t.description || '')
                    ].join(' ').toLowerCase();

                    if (searchText.includes(lowerQuery)) {
                        results.push({
                            api_name: data.api_name,
                            title: data.title,
                            version: data.version,
                            description: data.description,
                            cached_at: data.cached_at,
                            source: data.source,
                            tool_count: data.tools?.length || 0
                        });
                    }
                } catch {
                    // Skip files that can't be parsed
                    continue;
                }
            }

            return results;
        } catch (err) {
            console.warn(`Warning: Failed to search cache: ${err}`);
            return [];
        }
    }

    /**
     * List all cached API names.
     *
     * @returns List of cached API names
     */
    async listCachedApis(): Promise<string[]> {
        try {
            const files = await fs.readdir(this.cacheDir);
            return files
                .filter(f => f.endsWith('.json'))
                .map(f => f.replace('.json', ''));
        } catch (err) {
            console.warn(`Warning: Failed to list cached APIs: ${err}`);
            return [];
        }
    }

    /**
     * Clear cached API(s).
     *
     * @param name - API name to clear, or null to clear all
     * @returns True if cleared successfully
     */
    async clearCache(name?: string): Promise<boolean> {
        try {
            if (name) {
                // Clear specific API
                const cacheFile = path.join(this.cacheDir, `${name}.json`);
                await fs.unlink(cacheFile);
            } else {
                // Clear all APIs
                const files = await fs.readdir(this.cacheDir);
                for (const file of files) {
                    if (file.endsWith('.json')) {
                        await fs.unlink(path.join(this.cacheDir, file));
                    }
                }
            }
            return true;
        } catch (err) {
            console.warn(`Warning: Failed to clear cache: ${err}`);
            return false;
        }
    }

    // ==================== Session Persistence ====================

    /**
     * Save session context to disk.
     *
     * @param sessionId - Unique session identifier
     * @param context - AgentContext to save
     * @returns True if saved successfully
     */
    async saveSession(sessionId: string, context: AgentContext): Promise<boolean> {
        try {
            const sessionFile = path.join(this.sessionsDir, `${sessionId}.json`);

            const sessionData = {
                session_id: sessionId,
                saved_at: new Date().toISOString(),
                context: context.toDict()
            };

            await fs.writeFile(
                sessionFile,
                JSON.stringify(sessionData, null, 2),
                'utf-8'
            );

            return true;
        } catch (err) {
            console.warn(`Warning: Failed to save session '${sessionId}': ${err}`);
            return false;
        }
    }

    /**
     * Load session context from disk.
     *
     * @param sessionId - Session identifier
     * @returns AgentContext or null if not found
     */
    async loadSession(sessionId: string): Promise<AgentContext | null> {
        try {
            const sessionFile = path.join(this.sessionsDir, `${sessionId}.json`);
            const content = await fs.readFile(sessionFile, 'utf-8');
            const data = JSON.parse(content);

            return AgentContext.fromDict(data.context);
        } catch (err) {
            // File not found or parse error
            return null;
        }
    }

    /**
     * List all session metadata.
     *
     * @param limit - Maximum number of sessions to return (0 = all)
     * @returns List of session metadata, sorted by most recent first
     */
    async listSessions(limit: number = 0): Promise<Array<Record<string, any>>> {
        try {
            const files = await fs.readdir(this.sessionsDir);
            const sessions: Array<Record<string, any>> = [];

            for (const file of files) {
                if (!file.endsWith('.json')) continue;

                try {
                    const filePath = path.join(this.sessionsDir, file);
                    const content = await fs.readFile(filePath, 'utf-8');
                    const data = JSON.parse(content);

                    sessions.push({
                        session_id: data.session_id,
                        saved_at: data.saved_at,
                        agent_type: data.context?.agent_type,
                        user: data.context?.user,
                        workspace: data.context?.workspace
                    });
                } catch {
                    // Skip files that can't be parsed
                    continue;
                }
            }

            // Sort by saved_at descending (most recent first)
            sessions.sort((a, b) => {
                const dateA = new Date(a.saved_at).getTime();
                const dateB = new Date(b.saved_at).getTime();
                return dateB - dateA;
            });

            // Apply limit if specified
            if (limit > 0) {
                return sessions.slice(0, limit);
            }

            return sessions;
        } catch (err) {
            console.warn(`Warning: Failed to list sessions: ${err}`);
            return [];
        }
    }

    /**
     * Clean up old sessions, keeping only the most recent N.
     *
     * @param keepRecent - Number of recent sessions to keep
     * @returns Number of sessions deleted
     */
    async cleanupSessions(keepRecent: number = 10): Promise<number> {
        try {
            const sessions = await this.listSessions();

            if (sessions.length <= keepRecent) {
                return 0; // Nothing to clean up
            }

            const toDelete = sessions.slice(keepRecent);
            let deletedCount = 0;

            for (const session of toDelete) {
                try {
                    const sessionFile = path.join(
                        this.sessionsDir,
                        `${session.session_id}.json`
                    );
                    await fs.unlink(sessionFile);
                    deletedCount++;
                } catch {
                    // Continue even if deletion fails
                    continue;
                }
            }

            return deletedCount;
        } catch (err) {
            console.warn(`Warning: Failed to cleanup sessions: ${err}`);
            return 0;
        }
    }
}
